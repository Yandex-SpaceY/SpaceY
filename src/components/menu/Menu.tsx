import React, { FC, ReactElement } from 'react';
import { Link } from 'react-router-dom';
import cn from 'classnames';

import { GAME_NAME, ON_OFF_ITEMS } from 'constants/commonConstants';
import { MENU_ITEMS } from 'constants/menuConstants';

import './menu.scss';

export type TMenuItem = {
  title: string;
  route: string | null;
  action: string | null;
  titleAdditionIfModifierTrue?: string;
  titleAdditionIfModifierFalse?: string;
}

interface IMenu {
  menuItems?: TMenuItem[];
  isShown?: boolean;
  isWithTitle?: boolean;
  className?: string;
  soundModifier?: boolean;
  vibrationModifier?: boolean;
  handleAction?: (action: string) => void;
}

const Menu: FC<IMenu> = ({
  menuItems = MENU_ITEMS,
  isShown = true,
  isWithTitle = true,
  soundModifier = true,
  vibrationModifier = true,
  className,
  handleAction
}): ReactElement => {
  const getMenuItem = (title: string): ReactElement | null => {
    let result;
    switch (title) {
      case 'sound:': {
        result = soundModifier ? <span className='true'>{ON_OFF_ITEMS.ON}</span>
          : <span className='false'>{ON_OFF_ITEMS.OFF}</span>;
        break;
      }
      case 'vibration:': {
        result = vibrationModifier ? <span className='true'>{ON_OFF_ITEMS.ON}</span>
          : <span className='false'>{ON_OFF_ITEMS.OFF}</span>;
        break;
      }
      default: result = null;
    }

    return result;
  };

  return (
    <div className={cn('menu', className, (!isShown && 'hidden'))}>
      {isWithTitle
        && <h1>{GAME_NAME}</h1>
      }
      <div className='menu-items'>
        {
          menuItems.map(({ title, route, action }) => {
            const callback = () => {
              if (action && handleAction) {
                handleAction(action);
              }
            };

            if (route) {
              return (
                <Link key={title} className='menu-item' to={route} onClick={action ? callback : undefined}>
                  {title}
                </Link>
              );
            }

            if (action) {
              return (
                <span key={title} className='menu-item' onClick={callback}>
                  {title} {getMenuItem(title)}
                </span>
              );
            }

          })
        }
      </div>
    </div>

  );
};

export default Menu;
