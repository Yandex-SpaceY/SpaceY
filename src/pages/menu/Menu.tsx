import { MENU_ITEMS } from 'constants/commonConstants';
import React from 'react';
import './menu.scss';

const Signup = (): React.ReactElement => {
  return (
    <div className="menu">
      <div className="content-wrapper">
        <h1>SpaceY</h1>
        <div className="content">
          {MENU_ITEMS.map((item, i) => {
            const key = `${i}-${item.title}`;

            return (
              <h2 className="menu-item" key={key}>
                {item.title}
              </h2>
            );
          })}
        </div>
      </div>
      <h2 className="footer-item">DEBOARDING</h2>
    </div>
  );
};

export default Signup;
