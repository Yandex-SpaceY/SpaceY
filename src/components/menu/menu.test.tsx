import * as React from 'react';
import { shallow } from 'enzyme';

import Menu from './Menu';

it('check render component Menu', () => {
  const component = shallow(<Menu isTest={true} modifier={{ sound: true, vibration: true, theme: 'primary', skill: 1 }} />);

  expect(component).toMatchSnapshot();
});
