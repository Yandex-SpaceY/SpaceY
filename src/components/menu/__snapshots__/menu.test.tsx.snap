// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check render component Menu 1`] = `
<div
  className="menu"
>
  <h1>
    SpaceY
  </h1>
  <div
    className="menu-items"
  >
    <MenuItem
      isMobile={false}
      item={
        Object {
          "action": "GAME_START",
          "route": null,
          "title": "start",
        }
      }
      key="start"
      modifier={
        Object {
          "skill": 1,
          "sound": true,
          "theme": "primary",
          "vibration": true,
        }
      }
    />
    <MenuItem
      isMobile={false}
      item={
        Object {
          "action": "GAME_SKILL_SWITCH",
          "route": null,
          "title": "skill",
        }
      }
      key="skill"
      modifier={
        Object {
          "skill": 1,
          "sound": true,
          "theme": "primary",
          "vibration": true,
        }
      }
    />
    <MenuItem
      isMobile={false}
      item={
        Object {
          "action": "GAME_SOUND_SWITCH",
          "options": Object {
            "OFF": "OFF",
            "ON": "ON",
          },
          "route": null,
          "title": "sound",
        }
      }
      key="sound"
      modifier={
        Object {
          "skill": 1,
          "sound": true,
          "theme": "primary",
          "vibration": true,
        }
      }
    />
    <MenuItem
      isMobile={false}
      item={
        Object {
          "action": "GAME_VIBRATION_SWITCH",
          "mobileOnly": true,
          "options": Object {
            "OFF": "OFF",
            "ON": "ON",
          },
          "route": null,
          "title": "vibration",
        }
      }
      key="vibration"
      modifier={
        Object {
          "skill": 1,
          "sound": true,
          "theme": "primary",
          "vibration": true,
        }
      }
    />
    <MenuItem
      isMobile={false}
      item={
        Object {
          "action": null,
          "route": "/profile",
          "title": "my profile",
        }
      }
      key="my profile"
      modifier={
        Object {
          "skill": 1,
          "sound": true,
          "theme": "primary",
          "vibration": true,
        }
      }
    />
    <MenuItem
      isMobile={false}
      item={
        Object {
          "action": null,
          "route": "/leaderboard",
          "title": "best players",
        }
      }
      key="best players"
      modifier={
        Object {
          "skill": 1,
          "sound": true,
          "theme": "primary",
          "vibration": true,
        }
      }
    />
    <MenuItem
      isMobile={false}
      item={
        Object {
          "action": null,
          "route": "/forum",
          "title": "forum",
        }
      }
      key="forum"
      modifier={
        Object {
          "skill": 1,
          "sound": true,
          "theme": "primary",
          "vibration": true,
        }
      }
    />
    <MenuItem
      isMobile={false}
      item={
        Object {
          "action": "GAME_THEME_SWITCH",
          "route": null,
          "title": "theme",
        }
      }
      key="theme"
      modifier={
        Object {
          "skill": 1,
          "sound": true,
          "theme": "primary",
          "vibration": true,
        }
      }
    />
  </div>
</div>
`;
